### ISSUE TITLE
[REVIEW]: DRSP-Sim: A Simulator for Ride-Sharing with Pooling: Joint Matching, Pricing, Route Planning, and Dispatching

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3761]
**Submitting author:** @marina-haliem (<a href="http://orcid.org/0000-0002-9782-6591">Marina Haliem</a>)
**Repository:** https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator
**Branch with paper.md** (empty if default branch): 
**Version:** v1.1
**Editor:** @martinfleis
**Reviewers:** @RafalKucharskiPK, @martinfleis, @jGaboardi
**Archive:** Pending

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/4f1fc8e4cc30a2a620950b66cc07d5f7/status.svg)](https://joss.theoj.org/papers/4f1fc8e4cc30a2a620950b66cc07d5f7)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/4f1fc8e4cc30a2a620950b66cc07d5f7"><img src="https://joss.theoj.org/papers/4f1fc8e4cc30a2a620950b66cc07d5f7/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/4f1fc8e4cc30a2a620950b66cc07d5f7/status.svg)](https://joss.theoj.org/papers/4f1fc8e4cc30a2a620950b66cc07d5f7)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@RafalKucharskiPK & @martinfleis, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @martinfleis know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @RafalKucharskiPK
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@marina-haliem) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @martinfleis
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@marina-haliem) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @RafalKucharskiPK, @1tylermitchell it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1528

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/INFOCOM.2018.8485988 is OK
- 10.1109/TITS.2019.2931830 is OK
- 10.1073/pnas.1611675114 is OK
- 10.1016/j.cor.2016.07.020 is OK
- 10.1007/s12469-016-0139-6 is OK
- 10.1016/j.ejor.2012.05.028 is OK
- 10.1145/3385958.3430484 is OK
- 10.1109/TITS.2021.3096537 is OK
- 10.1109/TITS.2020.3048361 is OK
- 10.1109/tits.2021.3083740 is OK
- 10.1145/3408308.3431114 is OK
- 10.1007/s11116-019-10012-y is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1016/j.trc.2020.02.008 is INVALID because of 'https://doi.org/' prefix
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.10 s (470.3 files/s, 54457.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          42            753            958           3366
Markdown                         2             71              0            198
TeX                              1             16              0            194
YAML                             1             14              4             74
XML                              3              0              0             26
-------------------------------------------------------------------------------
SUM:                            49            854            962           3858
-------------------------------------------------------------------------------


Statistical information for the repository '191604a27c7b329dadbed4f1' was
gathered on 2021/09/23.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Marina Wagdy Wadea H            17         15361           1240          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Haliem, Marina Wagdy          4          100.0          4.8                0.00
Marina Wagdy Wadea H       5073           33.0          0.7               17.94
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03761/joss.03761/10.21105.joss.03761.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03761/joss.03761/10.21105.joss.03761.pdf) :page_facing_up: :point_left:

**Comment by martinfleis** [comment]
👋🏼 @marina-haliem @RafalKucharskiPK @1tylermitchell this is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `openjournals/joss-reviews#3761` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if any of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@martinfleis) if you have any questions/concerns.

**Comment by martinfleis** [comment]
I am copying here https://github.com/openjournals/joss-reviews/issues/3712#issuecomment-924912823 from @RafalKucharskiPK for completeness. 👇

Dear @marina-haliem 
 I cloned the repo, downloaded the data (from https://purr.purdue.edu/publications/3843/1), created conda env, installed most of dependencies and run `simulator_driver.py`.
 
 Minor issues that I managed to overcome:
1. you require pandas==0.22 - which is ancient - 2017(!), and it clashed with many of my other packages, I used lates pandas and so far it went OK.
2. The `db.sqlite3` path is your local path, I changed it to "data\db": 
![image](https://user-images.githubusercontent.com/20555451/134348441-ffc89cf0-d2f4-4f18-8d3c-735cb3fc9ba7.png)

3. The default simulation takes 8 days (in simulation) one day takes ca. one hour on my laptop - way too long for a quickstart, please provide a toy example for first users.

Nonetheless it did not work for me, after a single day the error occured. Then I changed to one day in config, and again after a day passed I receive:
```
Traceback (most recent call last):
  File "/Users/rkucharski/Documents/GitHub/Dynamic-RideSharing-Pooling-Simulator/simulator_driver.py", line 159, in <module>
    Sim_experiment.simulator.step()
  File "/Users/rkucharski/Documents/GitHub/Dynamic-RideSharing-Pooling-Simulator/simulator/simulator.py", line 71, in step
    if vehicle.agent_type == agent_codes.dqn_agent:
AttributeError: 'Vehicle' object has no attribute 'agent_type'

Process finished with exit code 1 
```
 so please create a stable version to reproduce.

PS. On the main github page I am missing quick walkthrough tutorial for first users, e.g. now I first see how to set 'OSRM' which happens to be not needed, but only later I see where is the data bundle to run simulations.

Good luck!

**Comment by marina-haliem** [comment]
Hi @RafalKucharskiPK,

Thank you for your comments. I have updated the requirements list, and the default db_dir path. Also, I've updated the Getting Started section, highlighting how to run a quick start of 1 day. Apologies regarding this error, there was a minor bug that I fixed. I have conducted some additional testing, and I believe it should run smoothly now for you.

Please, let us know if you come across any further issues.

Thank you,
Marina

**Comment by whedon** [comment]
:wave: @RafalKucharskiPK, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @1tylermitchell, please update us on how your review is going (this is an automated reminder).

**Comment by marina-haliem** [comment]
Thanks, @RafalKucharskiPK, for your comments.

- I am not able to replicate this issue. I've run this project so many times but never encountered a similar issue. The only thing is that the logs will need to be cleared in case of re-running the experiments, or renamed differently to distinguish different experiments using 'DEFAULT_LOG_DIR' in config/settings.py. If there is a specific error that you encountered, please let us know and we will try to resolve it.
- I've updated parse_results.py and log_analyzer.py to avoid the confusion of the parameters. By default, the parse_results.py will provide a tool to plot the results of the default log directory specified in 'DEFAULT_LOG_DIR'. Since the user can decide to compare any number of experiments without any constraints, it is left up to them to specify their relevant paths to generate relevant plots. I have also updated the relevant description in the README file for more clarity.

Thank you,
Marina

**Comment by martinfleis** [comment]
Hi @1tylermitchell @RafalKucharskiPK,

can you give me an indication of how much time do you need to finish the review (following the checklist above)? It's currently 8 weeks since the review started and it would be great if we could wrap this soon so @marina-haliem can work on the revisions.

Thanks!

**Comment by martinfleis** [comment]
@whedon remove @1tylermitchell as reviewer

**Comment by whedon** [comment]
OK, @1tylermitchell is no longer a reviewer

**Comment by RafalKucharskiPK** [comment]
Hi, 
sorry about that - I hope to give you feedback before christmas. I already run simulations and took a look at the text. 
PS. Do we wait for another reviewer or does it rely on my opinion only now?

**Comment by martinfleis** [comment]
Thanks, @RafalKucharskiPK!

> Do we wait for another reviewer or does it rely on my opinion only now?

I am looking for another reviewer at the moment.

**Comment by martinfleis** [comment]
@whedon add @martinfleis as reviewer

**Comment by whedon** [comment]
OK, @martinfleis is now a reviewer

**Comment by martinfleis** [comment]
@marina-haliem I have formally added myself as a second reviewer now (it is not a common solution but possible). I hope that it will help speeding up the review a bit as I don't expect that any new reviewer would be able to work on it before the upcoming holidays.

**Comment by RafalKucharskiPK** [comment]
in this case I also may commit to provide a review before end of the year

**Comment by martinfleis** [comment]
@RafalKucharskiPK That would be amazing! Thanks!

**Comment by martinfleis** [comment]
Hi @marina-haliem ,

First, apologies for the long review process. It is not always possible to make it swift, especially with more common unexpected issues these days.

I have started my review of DRSP-Sim and ticked relevant boxes in the checklist above. However, I cannot at the moment tick the rest, in some cases due to missing components of the submission, in others due to (I assume) technical limitations on my side.

**Installation**


- I am running an M1 mac laptop on which I am unable to install dependencies pinned in the `requirements.txt` file. My assumption is that it is caused by the chip architecture so I’ll try to do the same on an intel-based machine later.
- Due to this I wasn’t able to check Functionality and Performance yet.

**Example usage** and **Functionality documentation**

- I believe that the documentation (or a lack of it) is the major issue here. The ReadMe contains very rough set of instructions on how to run the script but there’s no other documentation.
- Functions of the package are not listed anywhere. You [need](https://joss.readthedocs.io/en/latest/review_criteria.html#api-documentation) API documentation.
- The functions are not documented even within the source code, so looking inside it doesn’t help. There are no docstrings.

**Automated tests**

- There are no [automated tests](https://joss.readthedocs.io/en/latest/review_criteria.html#tests) either. There is no way to objectively assess whether the software works as intended.

**Community guidelines**

- I have ticked community guidelines but I believe that it would benefit if you could expand it. There’s no developer documentation at the moment, so it is unclear how to create a dev environment and what needs to be done to prepare a PR (do I need to follow some code style, write tests, write documentation?).

Apart from these points from the checklist, I also have a couple of additional questions and a few minor comments.

- What was the role of your co-authors? There are no commits from them, can you explain their contribution?
- The links to the git repository in the Readme point to the old Purdue address. Can you update them to point to the current Github repository?
- Formatting of the code snippets in the Readme is broken in some cases (code is formatted as a standard text).
- How can I use the library to do a simulation of, say, Prague? What is the required data input? How to create it? All that should be documented if you want it to be reusable.
- If you want DRSP-Sim to be usable in the long term, you should avoid pinned versions of your dependencies. As @RafalKucharskiPK pointed above, your list breaks existing environments and doesn’t install on Arm-based machines.

Some of these points (documentation, tests) will undoubtedly require some work but you will need to address them to meet the [JOSS review criteria](https://joss.readthedocs.io/en/latest/review_criteria.html#).

Let me know if there’s anything unclear in my comments.

**Comment by RafalKucharskiPK** [comment]
Dear Editor, Dear Authors @marina-haliem ,

sorry for the delayed process and sorry that you did not manage to collect more reviewers.
I believe we may divide the review process into two separate parts:
1. Paper
2. Software

Since the software is well reviewed by @martinfleis and I had similar issues at the installation, I will focus more on the **paper**:

1. I am missing the use-cases. You well review the existing alternative models and highlight novelty, but I am still missing use-cases
2. The functionality of modules used in the architecture is missing. Why do we have them - how can we use them (parameterise) - and what are the use-cases for applying them in studies.
3. All in all the paper is very brief and does not give an overview of what is the software - it focuses a bit why is it better than other similar ones, but misses the point on what it really is.
4. The structure is misleading for me: you have Benchmarks as the section and then under Benchmarks you have details of pricing and mathing algorithms in subsections - are they benchmarks?

Please try to improve the clarity and completeness of the paper. By the way the issues are very similar to the issues found in the software description.

**Software**:
To summarize my previous issues - I am missing a walkthrough tutorial: starting from uses-case scenario (what do we want to simulate), through set-up, input data - through running and up to interpreation of the results. Without this I'd miss the whole story.

Hope this helps improve the paper!
Rafal Kucharski

**Comment by martinfleis** [comment]
Hi @marina-haliem,

Now that you have both reviews of the first round, please let us know how you plan to respond to the comments raised by reviewers and what is the expected timeline for that. Thanks!

Thanks @RafalKucharskiPK for your comments!

**Comment by martinfleis** [comment]
@RafalKucharskiPK Can you please update your checklist above?

**Comment by RafalKucharskiPK** [comment]
> @RafalKucharskiPK Can you please update your checklist above?

what do you mean exactly @martinfleis ?

**Comment by martinfleis** [comment]
@RafalKucharskiPK In the very top post in this issue is an automatically crated review checklist prepared for you. We use it to ensure that the paper fulfils all the JOSS requirements and wait until all boxes are ticked before accepting it. You should be able to edit that post and tick boxes you think the current version of the software and the paper meets. 
If you are unable to do so, I might need to re-invite you to join the JOSS org.

**Comment by marina-haliem** [comment]
Thank you for your comments and feedback.

**Reviewer 1** @martinfleis :

**Installation**
Would you elaborate on what issues are facing with this? I have personally run this project on a Mac laptop without any problems.

**Example usage and Functionality documentation**

1. I have documented the functionalities within the source code for more clarity as well as adding a thorough explanation of each module's functionalities in the paper under the (Simulator Architecture and Functionalities Section).
2. The modular code architecture diagram has also been updated to include all of the core functionalities of each sub-module.

**Testing**
For testing, we have provided manual steps in the Readme document that can be followed along with a sample expected output to check the expected functionality of the software.

**Community guidelines**
The developer guidelines are provided in the Readme including how to set up a dev environment, run the simulator, and generate plots to interpret the logs. No special code style is needed, just regular python code.

1. This codebase was initially developed over the old Purdue Github link, co-authors commits would have shown there. This Github link was mainly created for this submission. 
2. Repository link updated in the Readme. We intend to move the repository to another Github link in the future.
Formatting of the code snippets in the Readme is broken in some cases (code is formatted as a standard text).
3. This can be found under the (Data Generation) section in the Readme. If you would like to simulate any other city map (other than NY), you basically follow the same preprocessing steps listed there, but using the osm data of the required city as well as its corresponding trip data as well to generate the discretized city map.
4. I've updated parse_results.py and log_analyzer.py to avoid confusion of the parameters as @RafalKucharskiPK pointed out above. By default, the parse_results.py will provide a tool to plot the results of the default log directory specified in 'DEFAULT_LOG_DIR'. Since the user can decide to compare any number of experiments without any constraints, it is left up to them to specify their relevant paths to generate relevant plots. I have also updated the relevant description in the README file for more clarity.  I am not sure if Arm-based machines processors would cause any changes, but my team and I have successfully ran this project several times.

Thank you,
Marina

**Comment by marina-haliem** [comment]
Thank you for your comments and feedback.

Reviewer 2 @RafalKucharskiPK :

- For use cases, I relied on the Readme for that since the paper is limited in the number of pages. In the Readme, we have provided a thorough step-by-step explanation for simulator setup, input data, running the simulator, storing logs, and interpreting them by generating plots. Besides, anyone in the area can utilize the simulator for their own purposes, customizing the matching, pricing, and dispatching policies according to their research needs.
- For the functionality of modules used in the architecture, I have added a thorough explanation of each module's core functions in the paper under the (Simulator Architecture and Functionalities Section). The modular code architecture diagram has also been updated to include all of the core functionalities of each sub-module. For each module, our simulator provides one or two benchmarks for usage flexibility, but the simulator is meant to be utilized by any researcher in the area to plug in their own algorithms and compare the results.
- We emphasize the purpose of the simulator in the paper under the (Statement of need) section, stating that "Also, algorithms of customer-vehicle matching, route planning, pricing, and dispatching have been developed; however, these sub-problems tend to be studied separately and a complete integrated simulator that considers pooling is still lacking. We present the first RS simulator that integrates all of these together, provides a baseline for each and allows developers to plug-and-play algorithms to tackle any of the sub-problems while defaulting the rest to the provided benchmark. This, in turn, will greatly facilitate cross-study evaluations of any existing ridesharing algorithms, while providing guidance for future research."
- Under the (Benchmarks) section, we provide one or two benchmark algorithms for each sub-module (matching, pricing, and dispatching). This is what the simulator defaults to unless other algorithms are provided by the user for one or all of these sub-modules.

**Software:**
In the Readme, under the (Getting Stated) section, we provide a step-by-step explanation of creating a dev environment, setting up the simulator parameters that provide wide flexibility, loading and defining the input data, running the simulator, storing logs, post-processing and generating plots to interpret these logs. We also provide sample output of what plots would look like, if it were to compare two algorithms and if it were to visualize a summary of metrics for one algorithm.

Thank you,
Marina

**Comment by martinfleis** [comment]
Hi @marina-haliem, 

thank you for all the changes and a thorough response!

> Would you elaborate on what issues are facing with this? I have personally run this project on a Mac laptop without any problems.

Yes. When following the installation instructions, after cloning the repository, running

```
pip install -r requirements.txt
```
in a fresh environment fails with the following error message.

```
Collecting psycopg2-binary==2.7.4
  Using cached psycopg2-binary-2.7.4.tar.gz (426 kB)
  Preparing metadata (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /Users/martin/mambaforge/bin/python3.9 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-install-yvq80uz5/psycopg2-binary_d9f6f120662c49d7899e55e339bbb9ea/setup.py'"'"'; __file__='"'"'/private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-install-yvq80uz5/psycopg2-binary_d9f6f120662c49d7899e55e339bbb9ea/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-pip-egg-info-ku_fu6e_
       cwd: /private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-install-yvq80uz5/psycopg2-binary_d9f6f120662c49d7899e55e339bbb9ea/
  Complete output (23 lines):
  running egg_info
  creating /private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-pip-egg-info-ku_fu6e_/psycopg2_binary.egg-info
  writing /private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-pip-egg-info-ku_fu6e_/psycopg2_binary.egg-info/PKG-INFO
  writing dependency_links to /private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-pip-egg-info-ku_fu6e_/psycopg2_binary.egg-info/dependency_links.txt
  writing top-level names to /private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-pip-egg-info-ku_fu6e_/psycopg2_binary.egg-info/top_level.txt
  writing manifest file '/private/var/folders/2f/fhks6w_d0k556plcv3rfmshw0000gn/T/pip-pip-egg-info-ku_fu6e_/psycopg2_binary.egg-info/SOURCES.txt'
  
  Error: pg_config executable not found.
  
  pg_config is required to build psycopg2 from source.  Please add the directory
  containing pg_config to the $PATH or specify the full executable path with the
  option:
  
      python setup.py build_ext --pg-config /path/to/pg_config build ...
  
  or with the pg_config option in 'setup.cfg'.
  
  If you prefer to avoid building psycopg2 from source, please install the PyPI
  'psycopg2-binary' package instead.
  
  For further information please check the 'doc/src/install.rst' file (also at
  <http://initd.org/psycopg/docs/install.html>).
  
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/77/09/4991fcd9a8f4bea1ee3948e1729fa17c184d25bd10809bacc143626361b9/psycopg2-binary-2.7.4.tar.gz#sha256=de4f88f823037a71ea5ef3c1041d96b8a68d73343133edda684fd42f575bd9d7 (from https://pypi.org/simple/psycopg2-binary/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement psycopg2-binary==2.7.4 (from versions: 2.7.4, 2.7.5, 2.7.6, 2.7.6.1, 2.7.7, 2.8, 2.8.1, 2.8.2, 2.8.3, 2.8.4, 2.8.5, 2.8.6, 2.9, 2.9.1, 2.9.2, 2.9.3)
ERROR: No matching distribution found for psycopg2-binary==2.7.4
```

The version 2.7.4 you, for some reason, require is quite old (Feb 2018) and I believe that is causing the issue when you try to install it on Macs with M1 chips. There were no dependencies built for these chips back then.

**Comment by marina-haliem** [comment]
Hi @martinfleis ,
Thank you for the clarification. I have updated the requirements.txt. Hopefully, this fixes the problem, I have verified it is working as expected on different machines that I have access to.

Thank you,
Marina

**Comment by marina-haliem** [comment]
Hi @martinfleis ,
Any updates regarding this submission?

Thank you,
Marina

**Comment by martinfleis** [comment]
@marina-haliem It is waiting for the second round of revisions, sorry for delays.

@RafalKucharskiPK can you have a look at the response above and update the checklist in [the first post](https://github.com/openjournals/joss-reviews/issues/3761#issue-1005904121)?

**Comment by arfon** [comment]
:wave: @martinfleis – looks like this review has stalled somewhat?

**Comment by martinfleis** [comment]
Hey @RafalKucharskiPK, will you be able to check the response to your comments and update the checklist?

**Comment by martinfleis** [comment]
Hey @JGaboardi, sorry ti pick you up again :). It seems that we have lost a reviewer along the way here. Would you be willing to help out and review this submission? You will see that there already was one round of reviews done.

**Comment by jGaboardi** [comment]
@whedon generate pdf

**Comment by jGaboardi** [comment]
@editorialbot generate pdf

**Comment by martinfleis** [comment]
@editorialbot add @jGaboardi as reviewer

**Comment by jGaboardi** [comment]
@martinfleis, will the editorialbot generate a fresh checklist for me or shall I use the one from the previous reviewer?

**Comment by martinfleis** [comment]
Use `@editorialbot generate my checklist` command. I didn't realise this review is pre-editorialbot.

**Comment by jGaboardi** [comment]
@editorialbot generate my checklist

**Comment by jGaboardi** [comment]
@marina-haliem @martinfleis 

Before continuing with my review, I would like to raise two fairly large points (IMHO), that need to be addressed:

* **software**
  * One of the overarching goals of JOSS, and, similar outlets, is to advance the concept of (open source) software as a scientific artifact. While `DRSP-Sim` is technically open source, it is not accessible for generalized use (e.g. it can't be installed and is not available via `pip` or `conda-forge`). Moreover, as @martinfleis pointed out, there is no documentation site and there are no docstrings to speak of. Where [there are docstrings](https://github.com/marina-haliem/Dynamic-RideSharing-Pooling-Simulator/blob/master/common/geoutils.py#L30-L46), they are not formatted in a standard way.
* **paper**
  * The paper is well written, but 7 of the 17 references are to works involving authors of this manuscript. With a quick search there seems to be other related open software that should be cited. For example, [`MaaSSim`](https://github.com/RafalKucharskiPK/MaaSSim) by a previous reviewer.

**Comment by martinfleis** [comment]
@marina-haliem Can you respond to comments raised by @jGaboardi above? Please keep in mind that the review process of JOSS is different than in a standard journal and assumes direct communication rather than fixed rounds of review-response cycle. Thank you!

**Comment by martinfleis** [comment]
@marina-haliem Hi, I understand this review may not go as expected and I apologise for that. However, we need your cooperation at this stage to move forward, so I am just sending a gentle ping to respond to the comments above. Thank you!

**Comment by oliviaguest** [comment]
@martinfleis it might be wroth emailing @marina-haliem?

**Comment by martinfleis** [comment]
Hi @marina-haliem, I am really sorry but due to the lack of response, I will have to reject the submission. Last response is dated June 15, 2022 and neither messages here or emails were answered since then.

@oliviaguest Due to the lack of activity from the authors, I recommend rejecting this paper.

@RafalKucharskiPK, @jGaboardi Thank you for the time you spent reviewing this submission. I am sorry we are unable to finish the process towards acceptance and publishing this time.

**Comment by arfon** [comment]
@editorialbot reject

**Comment by arfon** [comment]
> Hi @marina-haliem, I am really sorry but due to the lack of response, I will have to reject the submission. Last response is dated June 15, 2022 and neither messages here or emails were answered since then.

Done. 

Thanks to all involved, sorry this one didn't work out.
